Mathlib/Data/Finsupp/MonomialOrder.lean

```diff
@@ -0,0 +1,411 @@
+is compatible with addition and for which `0` is the smallest element.
+Since several monomial orders may have to be used simultaneously, one cannot
+gets them as instances.
+In this formalization, they are presented as a structure `MonomialOrder` which encapsulates
+`MonomialOrder.toSyn`, an additive and monotone isomorphism to a linearly ordered cancellative
+additive commutative monoid.
+The entry `MonomialOrder.wf` asserts that `MonomialOrder.syn` is well founded.
+
+The terminology comes from commutative algebra and algebraic geometry, especially Gröbner bases,
+where `c : σ →₀ ℕ` are exponents of monomial.
+
+Given a monomial order `m : MonomialOrder σ`, we provide the notation
+`c ≼[m] d` and `c ≺[m] d` to compare `c d : σ →₀ ℕ` with respect to `m`.
+It is activated using `open scoped MonomialOrder`.
+
+## Examples
+
+We provide four examples of monomial orders, the most standard one in commutative algebra.
+
+* `MonomialOrder.lex` : the lexicographic ordering on `σ →₀ ℕ`.
+For this, `σ` needs to be embedded with an ordering relation which statisfies `WellFoundedGT σ`.
+(This last property is automatic when `σ` is finite).
+
+The type synonym is `Lex (σ →₀ ℕ)` and the two lemmas `MonomialOrder.lex_le_iff`
+and `MonomialOrder.lex_lt_iff` rewrite the ordering as comparisons in the type `Lex (σ →₀ ℕ)`.
+
+* `MonomialOrder.revLex` : the reverse lexicographic ordering on `σ →₀ ℕ`.
+For this, `σ` needs to be embedded with an ordering relation which statisfies `WellFoundedLT σ`.
+(This last property is automatic when `σ` is finite).
+
+The type synonym is `Lex (σᵒᵈ →₀ ℕ)` and the two lemmas `MonomialOrder.revLex_le_iff`
+and `MonomialOrder.revLex_lt_iff` rewrite the ordering as comparisons in the type `Lex (σᵒᵈ →₀ ℕ)`.
+
+* `MonomialOrder.degLex`: a variant of the lexicographic ordering that first compares degrees.
+For this, `σ` needs to be embedded with an ordering relation which statisfies `WellFoundedGT σ`.
+(This last property is automatic when `σ` is finite).
+
+The type synonym is `DegLex (σ →₀ ℕ)` and the two lemmas `MonomialOrder.degLex_le_iff`
+and `MonomialOrder.degLex_lt_iff` rewrite the ordering as comparisons in the type `Lex (σ →₀ ℕ)`.
+
+* `MonomialOrder.degRevLex`: a variant of the reverse lexicographic ordering
+that first compares degrees.
+For this, `σ` needs to be embedded with an ordering relation which statisfies `WellFoundedLT σ`.
+(This last property is automatic when `σ` is finite).
+
+The type synonym is `DegLex (σᵒᵈ →₀ ℕ)` and the two lemmas `MonomialOrder.degRevLex_le_iff`
+and `MonomialOrder.degRevLex_lt_iff` rewrite the ordering as comparisons in the type
+`DegLex (σᵒᵈ →₀ ℕ)`.
+
+-/
+
+/-- Monomial orders : equivalence of `σ →₀ ℕ` with a well ordered type -/
+structure MonomialOrder (σ : Type*) where
+  /-- The synonym type -/
+  syn : Type*
+  /-- `syn` is a linearly ordered cancellative additive commutative monoid -/
+  locacm : LinearOrderedCancelAddCommMonoid syn
```

Comment:
```suggestion
  locacm : LinearOrderedCancelAddCommMonoid syn := by infer_instance
```
Doing this, in the subsequent definition, you do not need to specify the field `locacm` if an instance can be found (which is the case in the definitions below). The same applies to `wf`.